// @tarquinen/opencode-dcp/dist/lib/ui/notification.js
import { countDistillationTokens, formatExtracted, formatPrunedItemsList, formatStatsHeader, formatTokenCount, formatProgressBar, } from "./utils";
export const PRUNE_REASON_LABELS = {
    completion: "Task Complete",
    noise: "Noise Removal",
    extraction: "Extraction",
};
function buildMinimalMessage(state, reason, distillation, showDistillation) {
    const extractedTokens = countDistillationTokens(distillation);
    const extractedSuffix = extractedTokens > 0 ? ` (distilled ${formatTokenCount(extractedTokens)})` : "";
    const reasonSuffix = reason && extractedTokens === 0 ? ` — ${PRUNE_REASON_LABELS[reason]}` : "";
    let message = formatStatsHeader(state.stats.totalPruneTokens, state.stats.pruneTokenCounter) +
        reasonSuffix +
        extractedSuffix;
    return message + formatExtracted(showDistillation ? distillation : undefined);
}
function buildDetailedMessage(state, reason, pruneToolIds, toolMetadata, workingDirectory, distillation, showDistillation) {
    let message = formatStatsHeader(state.stats.totalPruneTokens, state.stats.pruneTokenCounter);
    if (pruneToolIds.length > 0) {
        const pruneTokenCounterStr = `~${formatTokenCount(state.stats.pruneTokenCounter)}`;
        const extractedTokens = countDistillationTokens(distillation);
        const extractedSuffix = extractedTokens > 0 ? `, distilled ${formatTokenCount(extractedTokens)}` : "";
        const reasonLabel = reason && extractedTokens === 0 ? ` — ${PRUNE_REASON_LABELS[reason]}` : "";
        message += `\n\n▣ Pruning (${pruneTokenCounterStr}${extractedSuffix})${reasonLabel}`;
        const itemLines = formatPrunedItemsList(pruneToolIds, toolMetadata, workingDirectory);
        message += "\n" + itemLines.join("\n");
    }
    return (message + formatExtracted(showDistillation ? distillation : undefined)).trim();
}
const TOAST_BODY_MAX_LINES = 12;
const TOAST_SUMMARY_MAX_CHARS = 600;
function truncateToastBody(body, maxLines = TOAST_BODY_MAX_LINES) {
    const lines = body.split("\n");
    if (lines.length <= maxLines) {
        return body;
    }
    const kept = lines.slice(0, maxLines - 1);
    const remaining = lines.length - maxLines + 1;
    return kept.join("\n") + `\n... and ${remaining} more`;
}
function truncateToastSummary(summary, maxChars = TOAST_SUMMARY_MAX_CHARS) {
    if (summary.length <= maxChars) {
        return summary;
    }
    return summary.slice(0, maxChars - 3) + "...";
}
function truncateExtractedSection(message, maxChars = TOAST_SUMMARY_MAX_CHARS) {
    const marker = "\n\n▣ Extracted";
    const index = message.indexOf(marker);
    if (index === -1) {
        return message;
    }
    const extracted = message.slice(index);
    if (extracted.length <= maxChars) {
        return message;
    }
    return message.slice(0, index) + truncateToastSummary(extracted, maxChars);
}
export async function sendUnifiedNotification(client, logger, config, state, sessionId, pruneToolIds, toolMetadata, reason, params, workingDirectory, distillation) {
    const hasPruned = pruneToolIds.length > 0;
    if (!hasPruned) {
        return false;
    }
    if (config.pruneNotification === "off") {
        return false;
    }
    const showDistillation = config.tools.distill.showDistillation;
    const message = config.pruneNotification === "minimal"
        ? buildMinimalMessage(state, reason, distillation, showDistillation)
        : buildDetailedMessage(state, reason, pruneToolIds, toolMetadata, workingDirectory, distillation, showDistillation);
    if (config.pruneNotificationType === "toast") {
        let toastMessage = truncateExtractedSection(message);
        toastMessage =
            config.pruneNotification === "minimal" ? toastMessage : truncateToastBody(toastMessage);
        await client.tui.showToast({
            body: {
                title: "DCP: Prune Notification",
                message: toastMessage,
                variant: "info",
                duration: 5000,
            },
        });
        return true;
    }
    await sendIgnoredMessage(client, sessionId, message, params, logger);
    return true;
}
export async function sendCompressNotification(client, logger, config, state, sessionId, toolIds, messageIds, topic, summary, startResult, endResult, totalMessages, params) {
    if (config.pruneNotification === "off") {
        return false;
    }
    let message;
    if (config.pruneNotification === "minimal") {
        message = formatStatsHeader(state.stats.totalPruneTokens, state.stats.pruneTokenCounter);
    }
    else {
        message = formatStatsHeader(state.stats.totalPruneTokens, state.stats.pruneTokenCounter);
        const pruneTokenCounterStr = `~${formatTokenCount(state.stats.pruneTokenCounter)}`;
        const progressBar = formatProgressBar(totalMessages, startResult.messageIndex, endResult.messageIndex, 25);
        message += `\n\n▣ Compressing (${pruneTokenCounterStr}) ${progressBar}`;
        message += `\n→ Topic: ${topic}`;
        message += `\n→ Items: ${messageIds.length} messages`;
        if (toolIds.length > 0) {
            message += ` and ${toolIds.length} tools condensed`;
        }
        else {
            message += ` condensed`;
        }
        if (config.tools.compress.showCompression) {
            message += `\n→ Compression: ${summary}`;
        }
    }
    if (config.pruneNotificationType === "toast") {
        let toastMessage = message;
        if (config.tools.compress.showCompression) {
            const truncatedSummary = truncateToastSummary(summary);
            if (truncatedSummary !== summary) {
                toastMessage = toastMessage.replace(`\n→ Compression: ${summary}`, `\n→ Compression: ${truncatedSummary}`);
            }
        }
        toastMessage =
            config.pruneNotification === "minimal" ? toastMessage : truncateToastBody(toastMessage);
        await client.tui.showToast({
            body: {
                title: "DCP: Compress Notification",
                message: toastMessage,
                variant: "info",
                duration: 5000,
            },
        });
        return true;
    }
    await sendIgnoredMessage(client, sessionId, message, params, logger);
    return true;
}
export async function sendIgnoredMessage(client, sessionID, text, params, logger) {
    const agent = params.agent || undefined;
    const variant = params.variant || undefined;
    const model = params.providerId && params.modelId
        ? {
            providerID: params.providerId,
            modelID: params.modelId,
        }
        : undefined;
    try {
        await client.session.prompt({
            path: {
                id: sessionID,
            },
            body: {
                noReply: true,
                agent: agent,
                model: model,
                variant: variant,
                parts: [
                    {
                        type: "text",
                        text: text,
                        ignored: true,
                    },
                ],
            },
        });
    }
    catch (error) {
        logger.error("Failed to send notification", { error: error.message });
    }
}
