// @tarquinen/opencode-dcp/dist/lib/ui/utils.js
import { extractParameterKey } from "../messages/utils";
import { countTokens } from "../strategies/utils";
export function countDistillationTokens(distillation) {
    if (!distillation || distillation.length === 0)
        return 0;
    return countTokens(distillation.join("\n"));
}
export function formatExtracted(distillation) {
    if (!distillation || distillation.length === 0) {
        return "";
    }
    let result = `\n\n▣ Extracted`;
    for (const finding of distillation) {
        result += `\n───\n${finding}`;
    }
    return result;
}
export function formatStatsHeader(totalTokensSaved, pruneTokenCounter) {
    const totalTokensSavedStr = `~${formatTokenCount(totalTokensSaved + pruneTokenCounter)}`;
    return [`▣ DCP | ${totalTokensSavedStr} saved total`].join("\n");
}
export function formatTokenCount(tokens) {
    if (tokens >= 1000) {
        return `${(tokens / 1000).toFixed(1)}K`.replace(".0K", "K") + " tokens";
    }
    return tokens.toString() + " tokens";
}
export function truncate(str, maxLen = 60) {
    if (str.length <= maxLen)
        return str;
    return str.slice(0, maxLen - 3) + "...";
}
export function formatProgressBar(total, start, end, width = 20) {
    if (total <= 0)
        return `│${" ".repeat(width)}│`;
    const startIdx = Math.floor((start / total) * width);
    const endIdx = Math.min(width - 1, Math.floor((end / total) * width));
    let bar = "";
    for (let i = 0; i < width; i++) {
        if (i >= startIdx && i <= endIdx) {
            bar += "░";
        }
        else {
            bar += "█";
        }
    }
    return `│${bar}│`;
}
export function shortenPath(input, workingDirectory) {
    const inPathMatch = input.match(/^(.+) in (.+)$/);
    if (inPathMatch) {
        const prefix = inPathMatch[1];
        const pathPart = inPathMatch[2];
        const shortenedPath = shortenSinglePath(pathPart, workingDirectory);
        return `${prefix} in ${shortenedPath}`;
    }
    return shortenSinglePath(input, workingDirectory);
}
function shortenSinglePath(path, workingDirectory) {
    if (workingDirectory) {
        if (path.startsWith(workingDirectory + "/")) {
            return path.slice(workingDirectory.length + 1);
        }
        if (path === workingDirectory) {
            return ".";
        }
    }
    return path;
}
export function formatPrunedItemsList(pruneToolIds, toolMetadata, workingDirectory) {
    const lines = [];
    for (const id of pruneToolIds) {
        const metadata = toolMetadata.get(id);
        if (metadata) {
            const paramKey = extractParameterKey(metadata.tool, metadata.parameters);
            if (paramKey) {
                // Use 60 char limit to match notification style
                const displayKey = truncate(shortenPath(paramKey, workingDirectory), 60);
                lines.push(`→ ${metadata.tool}: ${displayKey}`);
            }
            else {
                lines.push(`→ ${metadata.tool}`);
            }
        }
    }
    const knownCount = pruneToolIds.filter((id) => toolMetadata.has(id)).length;
    const unknownCount = pruneToolIds.length - knownCount;
    if (unknownCount > 0) {
        lines.push(`→ (${unknownCount} tool${unknownCount > 1 ? "s" : ""} with unknown metadata)`);
    }
    return lines;
}
export function formatPruningResultForTool(prunedIds, toolMetadata, workingDirectory) {
    const lines = [];
    lines.push(`Context pruning complete. Pruned ${prunedIds.length} tool outputs.`);
    lines.push("");
    if (prunedIds.length > 0) {
        lines.push(`Semantically pruned (${prunedIds.length}):`);
        lines.push(...formatPrunedItemsList(prunedIds, toolMetadata, workingDirectory));
    }
    return lines.join("\n").trim();
}
