// @tarquinen/opencode-dcp/dist/lib/commands/help.js
/**
 * DCP Help command handler.
 * Shows available DCP commands and their descriptions.
 */
import { sendIgnoredMessage } from "../ui/notification";
import { getCurrentParams } from "../strategies/utils";
const BASE_COMMANDS = [
    ["/dcp context", "Show token usage breakdown for current session"],
    ["/dcp stats", "Show DCP pruning statistics"],
    ["/dcp sweep [n]", "Prune tools since last user message, or last n tools"],
    ["/dcp manual [on|off]", "Toggle manual mode or set explicit state"],
];
const TOOL_COMMANDS = {
    prune: ["/dcp prune [focus]", "Trigger manual prune tool execution"],
    distill: ["/dcp distill [focus]", "Trigger manual distill tool execution"],
    compress: ["/dcp compress [focus]", "Trigger manual compress tool execution"],
};
function getVisibleCommands(config) {
    const commands = [...BASE_COMMANDS];
    for (const tool of ["prune", "distill", "compress"]) {
        if (config.tools[tool].permission !== "deny") {
            commands.push(TOOL_COMMANDS[tool]);
        }
    }
    return commands;
}
function formatHelpMessage(manualMode, config) {
    const commands = getVisibleCommands(config);
    const colWidth = Math.max(...commands.map(([cmd]) => cmd.length)) + 4;
    const lines = [];
    lines.push("╭─────────────────────────────────────────────────────────────────────────╮");
    lines.push("│                              DCP Commands                               │");
    lines.push("╰─────────────────────────────────────────────────────────────────────────╯");
    lines.push("");
    lines.push(`  ${"Manual mode:".padEnd(colWidth)}${manualMode ? "ON" : "OFF"}`);
    lines.push("");
    for (const [cmd, desc] of commands) {
        lines.push(`  ${cmd.padEnd(colWidth)}${desc}`);
    }
    lines.push("");
    return lines.join("\n");
}
export async function handleHelpCommand(ctx) {
    const { client, state, logger, sessionId, messages } = ctx;
    const { config } = ctx;
    const message = formatHelpMessage(state.manualMode, config);
    const params = getCurrentParams(state, messages, logger);
    await sendIgnoredMessage(client, sessionId, message, params, logger);
    logger.info("Help command executed");
}
