// @tarquinen/opencode-dcp/dist/lib/state/state.js
import { loadSessionState } from "./persistence";
import { isSubAgentSession, findLastCompactionTimestamp, countTurns, resetOnCompaction, loadPruneMap, } from "./utils";
import { getLastUserMessage } from "../shared-utils";
export const checkSession = async (client, state, logger, messages, manualModeDefault) => {
    const lastUserMessage = getLastUserMessage(messages);
    if (!lastUserMessage) {
        return;
    }
    const lastSessionId = lastUserMessage.info.sessionID;
    if (state.sessionId === null || state.sessionId !== lastSessionId) {
        logger.info(`Session changed: ${state.sessionId} -> ${lastSessionId}`);
        try {
            await ensureSessionInitialized(client, state, lastSessionId, logger, messages, manualModeDefault);
        }
        catch (err) {
            logger.error("Failed to initialize session state", { error: err.message });
        }
    }
    const lastCompactionTimestamp = findLastCompactionTimestamp(messages);
    if (lastCompactionTimestamp > state.lastCompaction) {
        state.lastCompaction = lastCompactionTimestamp;
        resetOnCompaction(state);
        logger.info("Detected compaction - reset stale state", {
            timestamp: lastCompactionTimestamp,
        });
    }
    state.currentTurn = countTurns(state, messages);
};
export function createSessionState() {
    return {
        sessionId: null,
        isSubAgent: false,
        manualMode: false,
        pendingManualTrigger: null,
        prune: {
            tools: new Map(),
            messages: new Map(),
        },
        compressSummaries: [],
        stats: {
            pruneTokenCounter: 0,
            totalPruneTokens: 0,
        },
        toolParameters: new Map(),
        toolIdList: [],
        nudgeCounter: 0,
        lastToolPrune: false,
        lastCompaction: 0,
        currentTurn: 0,
        variant: undefined,
        modelContextLimit: undefined,
    };
}
export function resetSessionState(state) {
    state.sessionId = null;
    state.isSubAgent = false;
    state.manualMode = false;
    state.pendingManualTrigger = null;
    state.prune = {
        tools: new Map(),
        messages: new Map(),
    };
    state.compressSummaries = [];
    state.stats = {
        pruneTokenCounter: 0,
        totalPruneTokens: 0,
    };
    state.toolParameters.clear();
    state.toolIdList = [];
    state.nudgeCounter = 0;
    state.lastToolPrune = false;
    state.lastCompaction = 0;
    state.currentTurn = 0;
    state.variant = undefined;
    state.modelContextLimit = undefined;
}
export async function ensureSessionInitialized(client, state, sessionId, logger, messages, manualModeDefault) {
    if (state.sessionId === sessionId) {
        return;
    }
    // logger.info("session ID = " + sessionId)
    // logger.info("Initializing session state", { sessionId: sessionId })
    resetSessionState(state);
    state.manualMode = manualModeDefault;
    state.sessionId = sessionId;
    const isSubAgent = await isSubAgentSession(client, sessionId);
    state.isSubAgent = isSubAgent;
    // logger.info("isSubAgent = " + isSubAgent)
    state.lastCompaction = findLastCompactionTimestamp(messages);
    state.currentTurn = countTurns(state, messages);
    const persisted = await loadSessionState(sessionId, logger);
    if (persisted === null) {
        return;
    }
    state.prune.tools = loadPruneMap(persisted.prune.tools, persisted.prune.toolIds);
    state.prune.messages = loadPruneMap(persisted.prune.messages, persisted.prune.messageIds);
    state.compressSummaries = persisted.compressSummaries || [];
    state.stats = {
        pruneTokenCounter: persisted.stats?.pruneTokenCounter || 0,
        totalPruneTokens: persisted.stats?.totalPruneTokens || 0,
    };
}
